### ISSUE TITLE
DoRA implementation differs from PEFT DoRA

### ISSUE THREAD

**Issue by BenjaminBossan** [KohakuBlueleaf/LyCORIS#216]
@sayakpaul and I investigated an issue with loading a LyCORIS LoRA checkpoint which uses DoRA in diffusers. For some reason, we couldn't get the shapes of the DoRA scale vector to match with the shapes produced by PEFT (which is what diffusers uses under the hood). After some investigation, we think that the DoRA implementations diverge as the DoRA scale is applied along a different axis here compared to PEFT. To reproduce:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

from lycoris import create_lycoris, LycorisNetwork

class DemoNet(nn.Module):
    """
    A Simple Pytorch Module for demo
    """

    def __init__(self):
        super().__init__()
        self.test_1 = nn.Linear(784, 2048)
        self.te_2st = nn.Linear(2048, 784)
        self._3test = nn.Linear(784, 10)

    def forward(self, x):
        h = self.test_1(x)
        h = F.mish(h)
        h = self.te_2st(h)
        h = x + h
        h = self._3test(h)
        return h
	
# FIRST APPLY LYCORIS DORA
net = DemoNet()
LycorisNetwork.apply_preset({"target_name": ["test_1"]})
lycoris_net1 = create_lycoris(net, 1.0, linear_dim=16, linear_alpha=2.0, algo="lora", dora_wd=True)
lycoris_net1.apply_to()

print(lycoris_net1.lycoris_test_1.dora_scale.shape)
# prints: torch.Size([1, 784])

# NOW USE PEFT DORA
from peft import LoraConfig, get_peft_model
net = DemoNet()
lora_config = LoraConfig(target_modules=["test_1"], use_dora=True)
peft_model = get_peft_model(net, lora_config)

print(peft_model.base_model.model.test_1.lora_magnitude_vector["default"].weight.shape)
# prints: torch.Size([2048])
```

As we can see, LyCORIS applies DoRA along axis 0 and PEFT along axis 1. If this observation is correct, it would make the DoRA checkpoints incompatible between the two packages and would also mean that one of the two is implementing DoRA incorrectly.

**Comment by KohakuBlueleaf** [comment]
Linear(2048, 768) have weight matrix in shape [2048, 768]

based on the paper's description, they obtain weight decompose scale on "k" and their notation for weight matrix is "dxk"

![image](https://github.com/user-attachments/assets/ed3b5746-b642-4fd9-9519-2259191b7b9c)
![image](https://github.com/user-attachments/assets/960ff502-5221-4948-9812-56ed4be6a248)
![image](https://github.com/user-attachments/assets/fc5fafc2-d456-4395-82b6-cd615989877d)
![image](https://github.com/user-attachments/assets/226a3448-c7fe-4ee7-af4e-83ae14e36712)

**Comment by KohakuBlueleaf** [comment]
I think in your example, LyCORIS have correct implementation
BUT, in my personal opinion, I think both implementation will have similar result.

**Comment by BenjaminBossan** [comment]
Thanks for the quick response.

It can indeed be a bit confusing on what axis the DoRA scaling should be applied, especially with the transpose operation that's implicit in the linear layer. IIRC I had the wrong axis first when I started to [implement](https://github.com/huggingface/peft/pull/1474) it but one of DoRA authors reviewed the PR and made me aware of the mistake. Thus I'd like to assume it's correct in PEFT but maybe not :)

At the end of the day, it is not easy to make this change now, whether on the PEFT or LyCORIS side, as it would invalidate all existing DoRA checkpoints. So I guess we can just leave it as is if DoRA models are working fine in both implementations. It would just mean that checkpoints are incompatible between the two packages.

**Comment by KohakuBlueleaf** [comment]
> Thanks for the quick response.
> 
> It can indeed be a bit confusing on what axis the DoRA scaling should be applied, especially with the transpose operation that's implicit in the linear layer. IIRC I had the wrong axis first when I started to [implement](https://github.com/huggingface/peft/pull/1474) it but one of DoRA authors reviewed the PR and made me aware of the mistake. Thus I'd like to assume it's correct in PEFT but maybe not :)
> 
> At the end of the day, it is not easy to make this change now, whether on the PEFT or LyCORIS side, as it would invalidate all existing DoRA checkpoints. So I guess we can just leave it as is if DoRA models are working fine in both implementations. It would just mean that checkpoints are incompatible between the two packages.

Based on the equation in DoRA paper with W' = W+BA and B is [d, r] A is [r, k]
In Y = WX + B linear equation, it means k is for input axis and d is for output axis

Which is as same as how pytorch store their parameters (out_dim, in_dim)
I think LyCORIS is correct following the description in paper

But I won't deny the possibility that paper author actually want to apply on output axis

**Comment by KohakuBlueleaf** [comment]
> Thanks for the quick response.
> 
> It can indeed be a bit confusing on what axis the DoRA scaling should be applied, especially with the transpose operation that's implicit in the linear layer. IIRC I had the wrong axis first when I started to [implement](https://github.com/huggingface/peft/pull/1474) it but one of DoRA authors reviewed the PR and made me aware of the mistake. Thus I'd like to assume it's correct in PEFT but maybe not :)
> 
> At the end of the day, it is not easy to make this change now, whether on the PEFT or LyCORIS side, as it would invalidate all existing DoRA checkpoints. So I guess we can just leave it as is if DoRA models are working fine in both implementations. It would just mean that checkpoints are incompatible between the two packages.

If you think "apply on output axis" is somehow required
I can try to implement an option for user to select which to apply 

BUT, I may not break the default behaviour.

**Comment by KohakuBlueleaf** [comment]
I will support "loading" output axis ver of DoRA in LyCORIS first.

**Comment by BenjaminBossan** [comment]
I agree that breaking the existing method is not a good idea. Whether adding a new option to use the other axis is worth it, I don't know.

I dug through the original code by the DoRA authors (which now switched to PEFT) and found this line:

https://github.com/NVlabs/DoRA/blob/bd3ed6f91bcfb1a1725089b6ba0325ab79404de0/commonsense_reasoning/peft/src/peft/tuners/dora.py#L389

Here it looks like the vector has the shape of `out_features`, not `in_features`. So in the code example above, that corresponds to 2048, which is what we see for PEFT. If I understand that code correctly, it corresponds to the PEFT implementation. But I agree that this seems to contradict the use of `k` in the snippets that you cite.

**Comment by KohakuBlueleaf** [comment]
@BenjaminBossan ok I will take this as result of some misleading representation in paper
Will try to add some compatibility things in my side
not sure if PEFT will have it though. Since lot of user are using LyCORIS to training things too

**Comment by BenjaminBossan** [comment]
Maybe @nbasyl can comment on the notation and if it would make sense to have an option to swap the axis.

**Comment by KohakuBlueleaf** [comment]
@BenjaminBossan I added `wd_on_output` option in 3.1.1 and if you pass `wd_on_output=True` it will make correct weight decomposition

**Comment by catboxanon** [comment]
DoRAs trained with this flag don't work correctly in ComfyUI. In fact, as far as I'm aware, no UI even supports this flag currently. The code still performs calculations on the input rather than output.

This custom node would apply them directly.

```python
import torch
import folder_paths
import comfy.utils
import comfy.sd
import comfy.model_management
import comfy.lora

def dora_decompose(dora_scale, weight, lora_diff, alpha, strength, intermediate_dtype, function):
    dora_scale = comfy.model_management.cast_to_device(dora_scale, weight.device, intermediate_dtype)
    lora_diff *= alpha
    weight_calc = weight + function(lora_diff).type(weight.dtype)
    weight_norm = (
        weight_calc
        .reshape(weight_calc.shape[0], -1)
        .norm(dim=1, keepdim=True)
        .reshape(weight_calc.shape[0], *[1] * (weight_calc.dim() - 1))
    )

    weight_calc *= (dora_scale / weight_norm).type(weight.dtype)
    if strength != 1.0:
        weight_calc -= weight
        weight += strength * (weight_calc)
    else:
        weight[:] = weight_calc
    return weight

class ChangeDoRAType:
    @classmethod
    def INPUT_TYPES(s):
        return {
            "required": {
                "model": ("MODEL", {"tooltip": "The diffusion model the LoRA will be applied to."}),
                "wd_on_output": ("BOOLEAN", {"tooltip": "Whether to apply norm on output dims (True) or not (False)."}),
            }
        }

    RETURN_TYPES = ("MODEL",)
    OUTPUT_TOOLTIPS = ("The modified diffusion model.", "The modified CLIP model.")
    FUNCTION = "change_dora_type"

    CATEGORY = "loaders"
    DESCRIPTION = "LoRAs are used to modify diffusion and CLIP models, altering the way in which latents are denoised such as applying styles. Multiple LoRA nodes can be linked together."

    def change_dora_type(self, model, wd_on_output):
        if wd_on_output:
            if hasattr(comfy.lora, "original_weight_decompose"):
                comfy.lora.weight_decompose = dora_decompose
                return (model,)
            else:
                comfy.lora.original_weight_decompose = comfy.lora.weight_decompose
                comfy.lora.weight_decompose = dora_decompose
        else:
            if hasattr(comfy.lora, "original_weight_decompose"):
                comfy.lora.weight_decompose = comfy.lora.original_weight_decompose
        return (model,)

NODE_CLASS_MAPPINGS = {
    "ChangeDoRAType": ChangeDoRAType
}
```
